drivers/charger/charger_max20335.c

```diff
@@ -0,0 +1,270 @@
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/charger.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/linear_range.h>
+
+#define MAX20335_REG_STATUS_A 0x02
+#define MAX20335_REG_ILIMCNTL 0x09
+#define MAX20335_REG_CHG_CNTL_A 0x0A
+#define MAX20335_CHGCNTLA_BAT_REG_CFG_MASK GENMASK(4, 1)
+#define MAX20335_ILIMCNTL_MASK GENMASK(1, 0)
+#define MAX20335_STATUS_A_CHG_STAT_MASK GENMASK(2, 0)
+#define MAX20335_CHRG_EN_MASK BIT(0)
+#define MAX20335_CHRG_EN BIT(0)
+#define MAX20335_REG_CVC_VREG_MIN_UV 4050000U
+#define MAX20335_REG_CVC_VREG_STEP_UV 50000U
+#define MAX20335_REG_CVC_VREG_MIN_IDX 0x0U
+#define MAX20335_REG_CVC_VREG_MAX_IDX 0x0CU
+
+struct charger_max20335_config {
+	struct i2c_dt_spec bus;
+};
+
+enum {
+	MAX20335_CHARGER_OFF,
+	MAX20335_CHARGING_SUSPENDEND_DUE_TO_TEMPERATURE,
+	MAX20335_PRE_CHARGE_IN_PROGRESS,
+	MAX20335_FAST_CHARGE_IN_PROGRESS_1,
+	MAX20335_FAST_CHARGE_IN_PROGRESS_2,
+	MAX20335_MAINTAIN_CHARGE_IN_PROGRESS,
+	MAX20335_MAIN_CHARGER_TIMER_DONE,
+	MAX20335_CHARGER_FAULT_CONDITION,
+};
+
+static const struct linear_range charger_uv_range =
+	LINEAR_RANGE_INIT(MAX20335_REG_CVC_VREG_MIN_UV,
+			  MAX20335_REG_CVC_VREG_STEP_UV,
+			  MAX20335_REG_CVC_VREG_MIN_IDX,
+			  MAX20335_REG_CVC_VREG_MAX_IDX);
+
+static int charger_max20335_get_status(const struct device *dev, enum charger_status *status)
+{
+	const struct charger_max20335_config *const config = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->bus, MAX20335_REG_STATUS_A, &val);
+	if (ret) {
+		return ret;
+	}
+
+	val = FIELD_GET(MAX20335_STATUS_A_CHG_STAT_MASK, val);
+
+	switch (val) {
+	case MAX20335_CHARGER_OFF:
+		__fallthrough;
+	case MAX20335_CHARGING_SUSPENDEND_DUE_TO_TEMPERATURE:
+		__fallthrough;
+	case MAX20335_CHARGER_FAULT_CONDITION:
+		*status = CHARGER_STATUS_NOT_CHARGING;
+		break;
+	case MAX20335_PRE_CHARGE_IN_PROGRESS:
+		__fallthrough;
+	case MAX20335_FAST_CHARGE_IN_PROGRESS_1:
+		__fallthrough;
+	case MAX20335_FAST_CHARGE_IN_PROGRESS_2:
+		__fallthrough;
+	case MAX20335_MAINTAIN_CHARGE_IN_PROGRESS:
+		*status = CHARGER_STATUS_CHARGING;
+		break;
+	case MAX20335_MAIN_CHARGER_TIMER_DONE:
+		*status = CHARGER_STATUS_FULL;
+		break;
+	default:
+		*status = CHARGER_STATUS_UNKNOWN;
+		break;
+	};
+
+	return 0;
+}
+
+static int charger_max20335_set_constant_charge_voltage(const struct device *dev,
+							uint32_t voltage_uv)
+{
+	const struct charger_max20335_config *const config = dev->config;
+	uint16_t idx;
+	uint8_t val;
+	int ret;
+
+	ret = linear_range_get_index(&charger_uv_range, voltage_uv, &idx);
+	if (ret == -EINVAL) {
+		return ret;
+	}
+
+	val = FIELD_PREP(MAX20335_CHGCNTLA_BAT_REG_CFG_MASK, idx);
+
+	return i2c_reg_update_byte_dt(&config->bus,
+				      MAX20335_REG_CHG_CNTL_A,
+				      MAX20335_CHGCNTLA_BAT_REG_CFG_MASK,
+				      val);
+}
+
+static int charger_max20335_set_constant_charge_current(const struct device *dev,
```

Comment:
nit: You could remove `charger_max20335_` from the local functions.
(Style preference only though so feel free to ignore).

Reply:
Fixed.